MAINT: improve the accuracy of `np.sinc` for real arguments

person142 opened the pull request:
Since NumPy implements the normalized sinc function, argument
reduction can be performed in the `sin` factor without significant
loss of precision for large arguments. To keep the implementation
sane, `sinc` is reimplemented as a ufunc. For consistency the complex
version is also reimplemented as a ufunc, though the implementation
remains the same as before.

A couple of further notes:

- computing the `sin` factor is essentially the same as what one would do when implementing `sinpi`, compare to e.g. https://github.com/scipy/scipy/blob/master/scipy/special/_trig.pxd#L16
- since `cdiv` is declared `static` in `npy_math_complex.c.src` it isn't available in `funcs.inc.src`, so for now I copy-pasted the implementation into `funcs.inc.src`. It should probably be factored out into a header file somewhere, but I figured I would wait to solicit opinions on where.

person142 committed:
MAINT: improve the accuracy of `np.sinc` for real arguments

Since NumPy implements the normalized sinc function, argument
reduction can be performed in the `sin` factor without significant
loss of precision for large arguments. To keep the implementation
sane, `sinc` is reimplemented as a ufunc. For consistency the complex
version is also reimplemented as a ufunc, though the implementation
remains the same as before.

charris commented:
Enhancements like this are usually discussed on the mailing list first. There are a few other things to consider.

- A `sinc` function is needed in  `function_base.py` for backwards compatibility, especially if the 
  arguments have changed. The new function may need a new name.
- Ufunc documentation is in `/numpy/core/code_generators/ufunc_docstrings.py`. I'm a bit 
  curious if `_add_newdocs` actually worked for you.
- The function should possibly be implemented in the `npymath` library. Needs thinking about.

seberg commented:
> Enhancements like this are usually discussed on the mailing list first. There are a few other things to consider.
> 
> * A `sinc` function is needed in  `function_base.py` for backwards compatibility, especially if the
>   arguments have changed. The new function may need a new name.

The functions do not behave differently except for additional features which come with a ufunc, right (and possibly strange things for subclasses, but I think there moving to ufunc is rather an improvement)?

> * Ufunc documentation is in `/numpy/core/code_generators/ufunc_docstrings.py`. I'm a bit
>   curious if `_add_newdocs` actually worked for you.

This seems to be right? Or maybe there was a previous version I missed.

> * The function should possibly be implemented in the `npymath` library. Needs thinking about.

Something I have not worked with much, so I do no thave much of an opinion.

Closing and reopening to see if that makes tests pass. @person142 do you want to pick this up again? I can ping the list for you, on first sight making this a ufunc seems pretty reasonable to me, and the result is probably more reliable than before, although I am not s pecialist.

person142 commented:
@seberg I can pick it up again; pinging the list would be appreciated.

seberg commented:
@person142 I send a mail earlier. Seems nobody is opposed as such. Do you know if the precision gain for larger `x` is worth it? I first expected yes, but it seems a bit tricky to judge. Does this safe much precision compared to the already existing floating point errors for larger numbers (of course you are much better for exact integers). I could also imagine that there is a very nice speed improvement here.

If neither would be true, it may not be worth it. Especially, since this makes adding a `normalized=True` keyword (to switch to the other version) difficult (although that seems not a big priority).

person142 commented:
@seberg I responded on the mailing list-we can see what people think.

person142 commented:
Though specifically to the question of large `x`: here's an extreme example where the current argument reduction will be quite bad:

```
>>> x = 1e15 + 0.5
>>> (np.sinc(x) - mpmath.sincpi(x))/mpmath.sincpi(x)
```

On master I get:

```
mpf('-0.047440634063133452')
```

versus

```
mpf('0.0')
```

on this branch (a bit lucky: with the improved argument reduction we get half-integers exactly correct too, and then we happened to get correct rounding on the division).

seberg commented:
Yeah, I understand that this is an improvement in a sense, but I am right not sure how much it matters. If you assume that the input already has an error on the order of one ULP at least, that improvement would seem mostly meaningless, since you do not have exact integer values in any case.

seberg commented:
Ah, I now understand the improvement is specifically interesting close to 0? Still need to figure how considerable it is, but for now I assume it makes sense. The speed improvement is actually pretty small (maybe 20%), larger for very small arrays of course.
In any case, I am still somewhat in favor, because I think such a function should be a ufunc in any case, so it only needs a small improvement anyway to tip me over the edge (and it may be a larger improvement I guess).

seberg commented:
Sorry for the spam, but for the precision datapoint (relative precision):
```
xarr = np.linspace(0, 2.1, 200000)
res = []
for x in xarr:
    res.append((np.sinc(x) - mpmath.sincpi(x))/mpmath.sincpi(x))

res = np.asarray(res, dtype=object)
print(abs(res).mean(), abs(res).max())
```
Master:
```
1.70112344295248e-15 6.61977521930425e-11
```
This Branch:
```
5.64884361112036e-17 4.01208410359583e-16
```
So at least in relative terms, that does seem relevant to me?

person142 commented:
Generally speaking, on master the relative error is going to increase linearly as you approach a zero of `sinc`, while on this branch it will remain constant:

```python
import numpy as np
import mpmath
import matplotlib.pyplot as plt

n = np.arange(1, 15)
xarr = 2 + 0.1**n

res = []
for x in xarr:
    res.append(abs(float((np.sinc(x) - mpmath.sincpi(x))/mpmath.sincpi(x))))
res = np.asarray(res)

plt.plot(n, res)
plt.show()
```

which on master gives:

![sinc-error-master](https://user-images.githubusercontent.com/15134881/58376566-9e1fa380-7f22-11e9-84b6-12ca0ef57116.png)

and on this branch gives:

![sinc-error-branch](https://user-images.githubusercontent.com/15134881/58376580-c9a28e00-7f22-11e9-884d-094f683a69c9.png)

(A log-log plot would be clearer here were it not for the fact that relative error is sometimes zero on this branch.

person142 commented:
Anyhow, let me know if further characterization of the relative error on master versus this branch is desired or if we want to move forward. Otherwise feel free to close.

seberg commented:
No, I think we can move forward. I can have a pass over things if you like, just ping me when you think I should.

seberg commented on numpy/core/code_generators/generate_umath.py:
```diff
@@ -912,6 +912,12 @@ def english_upper(s):
+    Ufunc(1, 1, None,
+          docstrings.get('numpy.core.umath.sinc'),
+          None,
+          TD(inexact, f='sinc', astype={'e':'f'}),
```
I think the file does this commonly, so you can keep it, but a space after the `:` would be nicer.

seberg commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -557,6 +557,37 @@ NPY_INPLACE @type@ npy_cbrt@c@(@type@ x)
+	x = -x;
+    }
+
+    r = npy_fmod@c@(x, 2);
```
If you have any reference as to this method, I would be very happy about a comment in the function pointing to it.

person142 replied:
Hm various related implementations for `sinpi` exist:

- Go: https://github.com/golang/go/blob/50bd1c4d4eb4fac8ddeb5f063c099daccfb71b26/src/math/lgamma.go#L324

- Mpmath: https://github.com/fredrik-johansson/mpmath/blob/2858b1000ffdd8596defb50381dcb83de2bcccc6/mpmath/math2.py#L113

- Boost: https://github.com/boostorg/math/blob/c759981a2af05c7319c07e532e070e59a72cc74f/include/boost/math/special_functions/sin_pi.hpp#L23

- SciPy (which doesn't really count because I wrote it): https://github.com/scipy/scipy/blob/master/scipy/special/_trig.pxd#L16

Preferences for any of those? Unfortunately I don't know of a paper with a reference for `sinpi`-the technique is likely too minor to merit publication.

seberg commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -557,6 +557,37 @@ NPY_INPLACE @type@ npy_cbrt@c@(@type@ x)
+	return (@type@) NPY_NAN;
+    }
+    else if (npy_isinf(x)) {
+	return 0;
```
I guess these are necessary to avoid warnings in the `fmod`?

seberg commented on numpy/core/tests/test_umath.py:
```diff
@@ -1271,6 +1271,29 @@ def test_heaviside(self):
+        y = np.linspace(-5, 5, 100)
+        x, y = np.meshgrid(x, y)
+        z = x + 1j*y
+        assert_allclose(ncu.sinc(z), ncu.sin(np.pi*z)/(np.pi*z))
```
I wonder if we should add a test which checks for the improved relative precision compared to the previous method. Also we should have tests for all types, which is probably nicest by using pytest parameterization.

seberg reviewed (commented):
Just curious, I did not think about it, can similar things be done for the complex function? @charris had some old comments about it, can you have a look at these as well? Otherwise, just ping me when you have any updates (or feel I should just look more closely because it is almost done already).

person142 committed:
MAINT: fix spacing for `sinc` entry in `generate_umath`

person142 committed:
DOC: fix extraneous closing parenthesis in `sinc` docstring

person142 committed:
TST: improve tests for `sinc`

- Paremetrize the basic tests for real/complex to make sure all dtypes
  are tested
- Add a test for improved relative error around zeros of the function

person142 committed:
MAINT: add a `sinc` alias back to `lib/function_base.py`

person142 committed:
DOC: add a description of the argument reduction used in `sinc`

person142 commented:
Ok, pushed some commits to hopefully address the review comments. Re this comment

> A sinc function is needed in function_base.py for backwards compatibility, especially if the
arguments have changed. The new function may need a new name.

from @charris: I just imported the new `sinc` into `function_base.py`. Is that satisfactory?

With regards to:

> can similar things be done for the complex function?

Yes, absolutely. To have an absolutely correct implementation one should also take extra care to slightly stave off overflow for large imaginary arguments by dividing by the factor of `pi x` early when necessary.

Another remaining issue is what to do with the copy-pasted implementation of `cdiv` in `funcs.inc.src`. Since the other implementation is in `npy_math_complex.c.src` I guess that if it is decided that `sinc` should be included in `npymath` then the issue will go away. Otherwise I suppose I'll need to pull the implementations into a common header file somewhere.

rgommers commented:
> from @charris: I just imported the new `sinc` into `function_base.py`. Is that satisfactory?

`function_base` is not public, so if this is needed then raising a deprecation warning there would be fine. If we're making our code more convoluted to appease users of non-public modules, at least let's make sure we can clean it up later on.

charris commented:
@person142 Needs a rebase.

InessaPawson commented:
@person142 If you need help with rebasing this PR, please don't hesitate to let us know.

person142 commented:
I can handle the rebase, but I need to, er, remember that I need to do that at a time when I have access to a computer (gonna be a few weeks).

InessaPawson commented:
@person142 Please rebase to move this PR forward.

mattip commented:
@person142 ping.
